### ISSUE TITLE
ArrayIndexOutOfBoundsException with H264Encoder using YUV444

### ISSUE THREAD

**Issue by Intektor** [jcodec/jcodec#231]
I've been trying for a long time, and ok I might be new to JCodec, but I don't think this should be a thing, even If Im doing something wrong there should be a better message than:
Caused by: java.lang.ArrayIndexOutOfBoundsException: 108

This is my code: 

```
               H264Encoder encoder = new H264Encoder(new H264FixedRateControl(4));

                Picture picture = Picture.create(mc.displayWidth, mc.displayHeight, ColorSpace.YUV444);

                ByteBuffer buffer = ByteBuffer.allocate(mc.displayWidth * mc.displayHeight * 20);

                encoder.encodeFrame(picture, buffer);
```

So this happens to me every time I try to encode a default Picture, I don't think this should be a thing

java.lang.RuntimeException: java.lang.ArrayIndexOutOfBoundsException: 108
	at de.intektor.no_walls.vtransport.NWEntityRendererJ.updateCameraAndRender(NWEntityRendererJ.java:226)
	at net.minecraft.client.Minecraft.runGameLoop(Minecraft.java:1140)
	at net.minecraft.client.Minecraft.run(Minecraft.java:407)
	at net.minecraft.client.main.Main.main(Main.java:118)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at net.minecraftforge.gradle.GradleStartCommon.launch(GradleStartCommon.java:97)
	at GradleStart.main(GradleStart.java:26)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 108
	at org.jcodec.codecs.h264.io.CAVLC.writeACBlock(CAVLC.java:49)
	at org.jcodec.codecs.h264.encode.MBEncoderI16x16.writeAC(MBEncoderI16x16.java:157)
	at org.jcodec.codecs.h264.encode.MBEncoderI16x16.chromaResidual(MBEncoderI16x16.java:87)
	at org.jcodec.codecs.h264.encode.MBEncoderI16x16.chroma(MBEncoderI16x16.java:69)
	at org.jcodec.codecs.h264.encode.MBEncoderI16x16.encodeMacroblock(MBEncoderI16x16.java:49)
	at org.jcodec.codecs.h264.H264Encoder.encodeMacroblock(H264Encoder.java:295)
	at org.jcodec.codecs.h264.H264Encoder.encodeSlice(H264Encoder.java:276)
	at org.jcodec.codecs.h264.H264Encoder.doEncodeFrame(H264Encoder.java:167)
	at org.jcodec.codecs.h264.H264Encoder.encodeFrame(H264Encoder.java:103)
	at de.intektor.no_walls.vtransport.NWEntityRendererJ.updateCameraAndRender(NWEntityRendererJ.java:200)
	... 15 more

**Comment by svitvitskiy** [comment]
My first intuition is that it runs out buffer to write. Kind of weird because it looks like you are supplying a large enough buffer. Could you upload the image sequence you are trying to encode so I can reproduce?

**Comment by svitvitskiy** [comment]
And it looks like you are setting a very low qp, is it intentional to illustrate the issue with the exception? Try setting it to something like 20. Setting qp to 4 in combination with your Minecraft images (I deduced that from reading your stack trace) which have a lot of high freq content will blow up the compressed size because you are basically requesting jcodec to preserve all that detail at any cost. I am almost thinking to start looking at IPCM macroblocks. But you are right, jcodec should fail in a more civilized way.

**Comment by Intektor** [comment]
There is no image, I am creating a blank Picture using Picture#create, just copy the code and try out, it will fail

**Comment by zhuker** [comment]
@svitvitskiy i created a test case for the issue in this PR https://github.com/jcodec/jcodec/pull/233
it does fail like @Intektor describes when used with yuv444 even if the picture is blank (all zeros)

**Comment by mrieser** [comment]
This reminds me of my problem a while ago, see #173. H264Encoder seems to support only YUV420J, and you try to pass YUV444. I've made some changes to support YUV420, which I needed, see PR #174.

**Comment by svitvitskiy** [comment]
Fixed in PR #235 . The encoder now checks that the supplied image is yuv420j colorspace.
@mrieser, thank you for your PR and sorry for not reviewing it sooner. The h.264 encoder will only support the yuv420p (narrow) colorspace in the future. I still have to figure out the color situation there.
